index.js

```diff
@@ -747,17 +747,25 @@
 
   //  Array$chainRec :: ((a -> c, b -> c, a) -> Array c, a) -> Array b
   function Array$chainRec(f, x) {
-    var $todo = [x];
-    var $done = [];
-    while ($todo.length > 0) {
-      var xs = f(iterationNext, iterationDone, $todo.shift());
-      var $more = [];
-      for (var idx = 0; idx < xs.length; idx += 1) {
-        (xs[idx].done ? $done : $more).push(xs[idx].value);
+    var result = [];
+    var nil = {};
+    var xs = {head: x, tail: nil};
+    while (xs !== nil) {
+      var sx = nil;
+      f(iterationNext, iterationDone, xs.head).forEach(function(step) {
+        if (step.done) {
+          result.push(step.value);
+        } else {
+          sx = {head: step.value, tail: sx};
+        }
+      });
+      xs = xs.tail;
+      while (sx !== nil) {
+        xs = {head: sx.head, tail: xs};
+        sx = sx.tail;
       }
-      Array.prototype.unshift.apply($todo, $more);
```

Comment:
if `Function.apply` is the root cause of stack size, to support large array we maybe just need to check the limit of size arguments of function, slice it and apply it in loop, and according to https://stackoverflow.com/questions/22747068/is-there-a-max-number-of-arguments-javascript-functions-can-accept/22747272#22747272 the lowest limit is Chrome, with `0x10000` args.

but i am not sure how to do it though..

Reply:
That could work, but I'd prefer to stick with the linked list implementation (assuming it performs reasonably) as I find it easier to understand.